sno/pk_generation.py

```diff
@@ -109,16 +128,37 @@ def load_data_from_repo(self, repo):
             for pk, feature_hash in data["generatedPrimaryKeys"].items()
         }
 
-        # Hash of feature contents -> primary key(s) of imported feature(s), for every feature ever imported.
-        self.hash_to_pks = self._invert_pk_map(self.pk_to_hash)
-
-        # Subset of hash_to_pks - only contains primary keys that have not yet been assigned during the current import.
-        # Meaning, if we need a primary key for a feature, we should first check this dict to find a historical one that
-        # hasn't yet been assigned to a feature during this import.
-        self.hash_to_unassigned_pks = self._invert_pk_map(self.pk_to_hash)
-
-        # Next primary key to use if we can't find a historical but unassigned one in hash_to_unassigned_pks.
-        self.next_new_pk = max(self.pk_to_hash) + 1 if self.pk_to_hash else 1
+        # First primary key to use if we can't find a historical but unassigned one.
+        self.first_new_pk = max(self.pk_to_hash) + 1 if self.pk_to_hash else 1
+
+        # The number of inserts, deletes, previous- and current-feature-count, are related by the given formula:
+        # prev-FC + inserts - deletes = curr-FC
+        # Since we know prev-FC and curr-FC already, we can already calculate the number of inserts we can encounter
+        # before we know that (inserts + deletes) definitely exceeds the similarity_detection_limit, and once that many
+        # inserts are encountered, we give up on similarity detection.
+        feature_count_delta = self.feature_count - self.prev_dest_dataset.feature_count
+        self.similarity_detection_insert_limit = (
+            max(self.similarity_detection_limit - feature_count_delta, 0) // 2
+        )
```

Comment:
this formula doesn't seem quite right to me

suppose `similarity_detection_limit=100`, which means if `(deletes+inserts)` exceeds 100 we stop trying to do similarity detection

if feature_count is constant, ie `feature_count_delta=0`, it makes sense: `similarity_detection_insert_limit = 50`, ie if we see 51 inserts that are non-exact matches for previously-deleted things, then we know there are at least 51 deletes too (because we know the feature_count is constant so there must be the same number of deletes as inserts) and so we know we'll exceed our limit of 100 while trying to match them.

if feature_count increases (suppose `feature_count_delta=20`) then this formula calculates `similarity_detection_insert_limit=40`. When we see 41 inserts that are non-exact matches for previously-deleted things, we know there are at least 31 deletes too. However now we _haven't_ exceeded our limit of 100 (41+31=72)

Reply:
Oops - feature_count_delta should have been added, not subtracted - the formula then gives 60. When we see 61 inserts, we know there are at least 41 deletes too - 61 + 41 = 102.
I also had missed checking if abs(feature_delta) exceeds the similarity_detection_limit - the rearranged formula doesn't say anything about whether there can be a negative number of inserts or deletes - of course there can't be, but when given large feature_deltas the formula assumes there can. Math is hard. Anyway, added the check.